server/src/main/java/io/druid/query/extraction/LookupExtractionModule.java

```diff
@@ -0,0 +1,204 @@
+import io.druid.guice.annotations.Smile;
+import io.druid.initialization.DruidModule;
+import io.druid.server.DruidNode;
+import io.druid.server.initialization.ZkPathsConfig;
+import io.druid.server.listener.announcer.ListenerResourceAnnouncer;
+import io.druid.server.listener.announcer.ListeningAnnouncerConfig;
+import io.druid.server.listener.resource.AbstractListenerHandler;
+import io.druid.server.listener.resource.ListenerResource;
+import io.druid.server.namespace.cache.LookupCoordinatorManager;
+import org.apache.curator.utils.ZKPaths;
+
+import javax.ws.rs.Path;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+
+/**
+ * Include when a service can listen to lookups
+ */
+public class LookupExtractionModule implements DruidModule
+{
+  private static final String PROPERTY_BASE = "druid.zk.paths";
+
+  public static String getTierListenerPath(String tier)
+  {
+    return ZKPaths.makePath(LookupCoordinatorManager.LOOKUP_LISTEN_ANNOUNCE_KEY, tier);
+  }
+
+  @Override
+  public List<? extends Module> getJacksonModules()
+  {
+    return ImmutableList.of(new SimpleModule("LookupExtracitonModule-" + UUID.randomUUID().toString()));
+  }
+
+  @Override
+  public void configure(Binder binder)
+  {
+    JsonConfigProvider.bind(binder, PROPERTY_BASE, LookupListeningAnnouncerConfig.class);
+    binder.bind(LookupReferencesManager.class).in(ManageLifecycle.class);
+    binder.bind(LookupListeningResource.class).in(LazySingleton.class);
+    binder.bind(LookupResourceListenerAnnouncer.class).in(ManageLifecycle.class);
+    LifecycleModule.register(binder, LookupReferencesManager.class);
+    LifecycleModule.register(binder, LookupListeningResource.class);
+    LifecycleModule.register(binder, LookupResourceListenerAnnouncer.class);
+    Jerseys.addResource(binder, LookupListeningResource.class);
+  }
+}
+
+@Path(ListenerResource.BASE_PATH + "/" + LookupCoordinatorManager.LOOKUP_LISTEN_ANNOUNCE_KEY)
+class LookupListeningResource extends ListenerResource
+{
+  private static final Logger LOG = new Logger(LookupListeningResource.class);
+
+  @Inject
+  public LookupListeningResource(
+      final @Json ObjectMapper jsonMapper,
+      final @Smile ObjectMapper smileMapper,
+      final LookupReferencesManager manager
+  )
+  {
+    super(
+        jsonMapper,
+        smileMapper,
+        new AbstractListenerHandler<LookupExtractorFactory>(new TypeReference<LookupExtractorFactory>()
+        {
+        })
+        {
+          @Override
+          public synchronized Object post(final Map<String, LookupExtractorFactory> lookups)
+              throws Exception
+          {
+            final Map<String, LookupExtractorFactory> failedUpdates = new HashMap<>();
+            for (final String name : lookups.keySet()) {
+              final LookupExtractorFactory factory = lookups.get(name);
+              try {
+                if (!manager.updateIfNew(name, factory)) {
```

Comment:
where race condition means that a node is actively having queries against it, and I want to be able to continue issuing those queries.